### PR TITLE
[None][perf] fuse DSpark attention and RMSNorm RoPE

### PR DESCRIPTION
**Pull request by longlee0622**
## What changed

- Add a Blackwell CuteDSL custom op that fuses DSpark rolling-KV update/read, validity handling, QK, attention-sink online softmax, and PV.
- Add a second CuteDSL custom op that fuses the DSpark RMSNorm and adjacent-pair RoPE paths, including inverse RoPE and RMSNorm-only modes.
- Integrate both ops into the batched DSpark attention path with the existing PyTorch implementation retained as a fallback.
- Compile with symbolic batch/row dimensions so normal warmup produces kernels reusable across inference batch sizes.

## Why

The prior batched implementation materialized top-k indices, gathered/concatenated KV, scores, probabilities, and several RMSNorm/RoPE intermediates. Nsight Systems showed this as many scatter, copy, cat, divide, mask, reduction, and elementwise kernels in the DSpark critical path.

The fused attention kernel directly addresses the fixed sliding-window plus current-block layout and avoids those intermediates. The RMSNorm/RoPE kernel replaces repeated FP32 materialization and complex-tensor operations with one vectorized kernel per transform.

## Impact

On an internal B200 microbenchmark:

- fused sparse attention: 67.5 us vs 312.3 us (4.63x faster)
- five RMSNorm/RoPE transforms: 295.5 us vs 732.2 us (2.48x faster)

Unsupported devices, dtypes, shapes, or layouts continue to use the existing implementation. The optimized path is currently gated to the existing SM100-family check (SM100/SM103).

## Validation

- `pre-commit run --files <7 changed files>`
- `pytest -q tests/unittest/_torch/speculative/test_dspark_cute_dsl_attention.py tests/unittest/_torch/speculative/test_dspark_cute_dsl_rmsnorm_rope.py -s` (10 passed on B200)
- `pytest -q tests/unittest/_torch/speculative/hw_agnostic/test_dspark_attention.py tests/unittest/_torch/speculative/hw_agnostic/test_dspark_cuda_graph.py -s` (35 passed)

The GPU tests cover reference parity, production-shape end-to-end parity, strided rolling cache updates, CUDA Graph replay, all fused RMSNorm/RoPE modes, and compile-cache reuse across batch sizes.

## Dev Engineer Review

- Added fused Blackwell CuteDSL kernels for DSpark attention and RMSNorm/RoPE.
- Added SM100/SM103 gating, input validation, compilation caching, symbolic batch dimensions, and fake implementations.
- Integrated the fused operations with PyTorch fallbacks.
- Preserved rolling KV updates, attention sinks, inverse RoPE, and RMSNorm-only modes.
- Added B200 CI entries for both fused operations.
- Reported benchmarks show 4.63x faster fused attention and 2.48x faster RMSNorm/RoPE processing.
- Main review risk remains kernel correctness across layouts, dimensions, and cache states. Parity, cache, graph replay, and validation tests reduce this risk.

## QA Engineer Review

Added tests for:

- DSpark attention support gates, validation, reference parity, cache updates, CUDA Graph replay, compilation reuse, and fused-versus-fallback equivalence.
- DSpark RMSNorm/RoPE support gates, validation, reference parity, CUDA Graph replay, and compilation reuse.
- CPU fallback weighting, unsupported-architecture handling, and cache-dimension validation.

The modified test-list file is `tests/integration/test_lists/test-db/l0_b200.yml`. It adds CI entries for DSpark CuteDSL attention and RMSNorm/RoPE coverage.

The GPU tests are covered by the B200 CI entries. The CPU tests are not represented by the reported test-list entries.

Verdict: sufficient.

### CODE REVIEW

File: tensorrt_llm/_torch/models/dspark/attention.py [hunk 1]
```diff
@@ -282,6 +294,45 @@ def _rope_last_dims_batched(
...
+        )
+
+    if apply_rmsnorm:
+        if apply_weight:
+            t = _rmsnorm(t, weight, eps)
+        else:
+            t = t * torch.rsqrt(t.square().mean(-1, keepdim=True) + eps)
+    if rope_head_dim > 0:
+        t = _rope_last_dims_batched(t, rope_head_dim, freqs_cis, inverse=inverse_rope)
+    return t
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/models/dspark/attention.py:340]: +1

File: tensorrt_llm/_torch/models/dspark/attention.py [hunk 2]
```diff
@@ -282,6 +294,46 @@ def _rope_last_dims_batched(
...
+                apply_weight,
+                apply_rmsnorm,
+                inverse_rope,
+            )
+
+    if apply_rmsnorm:
+        if apply_weight:
+            t = _rmsnorm(t, weight, eps)
+        else:
+            t = t * torch.rsqrt(t.square().mean(-1, keepdim=True) + eps)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/dspark/attention.py:335]: This branch computes the RMS in bf16 while the fused kernel reduces and scales in fp32. It is the path taken by the per-head query RMS at line 493, so on SM100 this PR changes production numerics rather than being a bit-neutral optimization - likely also why the parity test needs `rtol=8e-2`. Could you confirm draft acceptance rate is unaffected, and consider making this branch use `.float()` like `_rmsnorm` does so the fallback and the kernel agree?

File: tensorrt_llm/_torch/models/dspark/attention.py [hunk 3]
```diff
@@ -422,38 +474,77 @@ def dspark_attention_forward_batched(
...
     write_target = kv_cache if persist else kv_cache.clone()
-    slot_pos = start_pos % window_size  # [G]
-    write_target[slots, slot_pos] = main_kv.squeeze(1).to(write_target.dtype)
-    cache_rows = write_target[slots]  # [G, window, head_dim]
-    kv_full = torch.cat([cache_rows, kv], dim=1)  # [G, window + block, head_dim]
-    topk = get_dspark_topk_idxs_batched(window_size, block, start_pos)
-    o = dspark_sparse_attn(q, kv_full, attn_sink, topk, softmax_scale)  # [G, block, h, head_dim]
-    o = _rope_last_dims_batched(o, rd, blk_freqs, inverse=True)
+    main_kv_flat = main_kv.squeeze(1).to(write_target.dtype)
+    if IS_CUTLASS_DSL_AVAILABLE and is_fused_dspark_attention_supported(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/dspark/attention.py]: When the support predicate rejects the inputs the code falls back silently, so a dtype or shape drift that disables this 4.6x path would surface only as an unexplained perf regression. Consider logging the specific failing condition once (`logger.debug`, or a warn-once) from the predicate so the fallback is observable.

File: tensorrt_llm/_torch/models/dspark/attention.py [hunk 4]
```diff
@@ -282,6 +294,46 @@ def _rope_last_dims_batched(
...
+    freqs_cis: torch.Tensor,
+    *,
+    num_heads: int = 1,
+    apply_weight: bool = True,
+    apply_rmsnorm: bool = True,
+    inverse_rope: bool = False,
+) -> torch.Tensor:
+    """Fuse DSpark RMSNorm and last-dimension RoPE when supported."""
+    if IS_CUTLASS_DSL_AVAILABLE:
+        freqs_real = torch.view_as_real(freqs_cis).reshape(-1, freqs_cis.shape[-1], 2)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/models/dspark/attention.py:317]: `IS_CUTLASS_DSL_AVAILABLE` is purely an import check with no architecture gating, so on a non-SM100 GPU with cutlass DSL installed this builds `freqs_real` on every call (5x per forward) only to be rejected by `is_sm_100f()` on the first line of the predicate. Hoisting the (lru_cached) `is_sm_100f()` check above this line keeps that off the host path.

**Review by BowenFu — APPROVED** [review]

**Review by yingguo-trt — APPROVED** [review]

**Review by litaotju — APPROVED** [review]
